app/src/behaviors/behavior_hold_tap.c

```diff
@@ -0,0 +1,491 @@
+/*
+ * Copyright (c) 2020 Cody McGinnis, Okke Formsma
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#define DT_DRV_COMPAT zmk_behavior_hold_tap
+
+#include <device.h>
+#include <drivers/behavior.h>
+#include <logging/log.h>
+#include <zmk/behavior.h>
+
+#include <zmk/matrix.h>
+#include <zmk/endpoints.h>
+#include <zmk/event-manager.h>
+#include <zmk/events/position-state-changed.h>
+#include <zmk/events/keycode-state-changed.h>
+#include <zmk/events/modifiers-state-changed.h>
+#include <zmk/hid.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#if DT_NODE_EXISTS(DT_DRV_INST(0))
+
+/************************************************************ DATA SETUP */
+#define ZMK_BHV_HOLD_TAP_MAX_HELD 10
+#define ZMK_BHV_HOLD_TAP_MAX_CAPTURED_EVENTS 40
+
+// increase if you have keyboard with more keys.
+#define ZMK_BHV_HOLD_TAP_POSITION_NOT_USED 9999
+
+struct behavior_hold_tap_behaviors {
+	struct zmk_behavior_binding tap;
+	struct zmk_behavior_binding hold;
+};
+
+typedef k_timeout_t (*timer_func)();
+
+struct behavior_hold_tap_config {
+	timer_func tapping_term_ms;
+	struct behavior_hold_tap_behaviors *behaviors;
+	char *flavor;
+};
+
+// this data is specific for each hold-tap
+struct active_hold_tap {
+	s32_t position;
+	u32_t param_hold;
+	u32_t param_tap;
+	bool is_decided;
+	bool is_hold;
+	const struct behavior_hold_tap_config *config;
+	struct k_delayed_work work;
+	bool work_is_cancelled;
+};
+
+// The undecided hold tap is the hold tap that needs to be decided before
+// other keypress events can be released. While the undecided_hold_tap is
+// not NULL, most events are captured in captured_events.
+// After the hold_tap is decided, it will stay in the active_hold_taps until
+// its key-up has been processed and the delayed work is cleaned up.
+struct active_hold_tap *undecided_hold_tap = NULL;
+struct active_hold_tap active_hold_taps[ZMK_BHV_HOLD_TAP_MAX_HELD] = {};
+// We capture most position_state_changed events and some modifiers_state_changed events.
+const struct zmk_event_header *captured_events[ZMK_BHV_HOLD_TAP_MAX_CAPTURED_EVENTS] = {};
+
+/************************************************************ CAPTURED POSITION HELPER FUNCTIONS */
+static int capture_event(const struct zmk_event_header *event)
+{
+	for (int i = 0; i < ZMK_BHV_HOLD_TAP_MAX_CAPTURED_EVENTS; i++) {
+		if (captured_events[i] == NULL) {
+			captured_events[i] = event;
+			return 0;
+		}
+	}
+	return -ENOMEM;
+}
+
+static struct position_state_changed *find_captured_keydown_event(u32_t position)
+{
+	struct position_state_changed *last_match = NULL;
+	for (int i = 0; i < ZMK_BHV_HOLD_TAP_MAX_CAPTURED_EVENTS; i++) {
+		const struct zmk_event_header *eh = captured_events[i];
+		if (eh == NULL) {
```

Comment:
So... this makes sense, but *assumes* future refactors don't end up with a "sparse" `captured_events` array. Is this a reasonable assumption? If so, fine, just thought I'd ask the question.

Reply:
I think it is reasonable. A captured event should not be released or deleted on it's own, all decision about queued events should be made by the listener.